help_to_heat/frontdoor/views.py

```diff
@@ -238,26 +302,43 @@ def reset_epc_details(session_id):
 
 
 class PageView(utils.MethodDispatcher):
-    def get(self, request, session_id, page_name, data=None, errors=None, is_change_page=False):
+    def get(self, request, session_id, page_name, extra_data=None, errors=None, is_change_page=False):
         if not errors:
             errors = {}
-        if not data:
-            data = interface.api.session.get_answer(session_id, page_name)
+        answers = interface.api.session.get_session(session_id)
+        data = interface.api.session.get_answer(session_id, page_name)
+        if extra_data:
+            data = {**data, **extra_data}
+        click_choice = request.GET.get("click")
+        if click_choice is not None:
+            data = self.save_click_data(data, session_id, page_name, click_choice)
+            interface.api.session.save_answer(session_id, page_name, data)
+
+            # add new data to full answers object to calculate new next page
+            answers = {**answers, **data}
+            next_page_name = get_next_page(page_name, answers)
+            if next_page_name != unknown_page:
+                return redirect(page_name_to_url(session_id, next_page_name))
+            else:
+                return redirect("/sorry")
```

Comment:
I agree I'm not super happy with it, though I'm not too sure of alternate ways to do it.

the difficulty comes from that this arises from three different possible routes that lead to enter manually page:
- `address` -> enter manually
- `epc-select` -> enter manually
- `address-select` -> enter manually

this means that the address input page has three possible previous pages, which makes hard coding in the previous page difficult. right now we can discern which one is the correct one since there is an answer saved to the session when you click enter manually

Reply:
thinking a bit further it could be possible to solve this by registering each manual address page as a different URL, which can be done I think by adding multiple register page decorators to one pageview subclass. then it'd be possible to have the links go to detectably different places

Reply:
We could do something like that, or can we know which page we're going back to based on what data we've already got? i.e.
* if there's no address data, go back to `address`
* if there's address data and country isn't Scotland, go back to `epc-select`
* if there's address data and country is Scotland, go back to `address-select`

Appreciate this is duplicating some of the forward-navigation logic, but I think it's the lesser of all evils here, and this is a sufficiently unique case that we can justify it and add comments to make it safe.

Reply:
I think post our chat I'll look into whether these can be separate pages for each case, since:
- it'll avoid routing based on query params
- we can save an answer on post (after entering manually) to record where the user decided to enter manually. most of our answers are saved on post so this remains consistent
- the back link can be guessed easily if the URLs differ